### ISSUE TITLE
Add MVP OAuth 2.0 Support

### ISSUE THREAD

**Issue by mrjones-plip** [medic/cht-core#8849]
**Is your feature request related to a problem? Please describe.**
When deployments need to add a feature to the CHT, or want to host a related services like an external dashboard, one of the main challenges is that there is no way to authenticate against CHT itself.  

**Describe the solution you'd like**
It would be great if the CHT could implement [OAuth 2.0](https://oauth.net/2/) to enable remote systems to use the CHT for credentials. 

**Describe alternatives you've considered**
As a work around today,  systems can ask the user to enter their CHT Credentials anew on their own site and proxy over their requests as the [User Man. Tool does today](https://github.com/medic/cht-user-management/).

**Additional context**
This should be a minimum viable product (MVP) effort to support a "plugin" of sorts that allows external developers to specifically build the  [User Man. Tool does today](https://github.com/medic/cht-user-management/). This means if there's other parts of the OAuth 2.0 spec that take more time to implement but will not be used right, we shouldn't do that part.

**Comment by jkuester** [comment]
Just adding @m5r's [Slack update](https://medic.slack.com/archives/C02FMC810UE/p1707398576152059) here for posterity sake!  In our discussion today, we went over 2 OAuth2 server implementations that we could add to the CHT (either as a standalone server or existing alongside the CHT API. The OAuth spec being confusing and time-consuming to implement properly, it makes more sense to depend on a reliable and well-maintained third-party package:

- https://github.com/node-oauth/express-oauth-server
- https://github.com/panva/node-oidc-provider

Implementing the OAuth flow should be relatively straightforward but it's not completely clear yet how we're going to make CouchDB accept the access tokens we'll hand out to third parties to authenticate their users, potentially acting as middlemen to map access tokens with CouchDB sessions to authorize requests.  (One thought from @sugat009 was that perhaps we could just repurpose the session cookie from Couch somehow as the OAuth token given back to clients so that the CHT server could be more of a pass-through and not have to worry about mapping sessions.  Almost certainly not what was intended by the OAuth spec, but might radically simplify things. :shrug: )

The next steps are to further evaluate these packages and how we could potentially integrate them with the CHT and with CouchDB.  Then we will reconvene on Monday to decide and plan the next steps

**Comment by mrjones-plip** [comment]
Awesome - exactly what this ticket was meant for - thanks @jkuester !  

BTW - I had the same thought as @sugat009 for inappropriate reuse of session cookie :sweat_smile: )

**Comment by garethbowen** [comment]
"inappropriate reuse" is a very interesting idea. I can't think of a reason why it's a bad idea...

Another option is to remove authentication from CouchDB altogether and use [proxy authentication](https://docs.couchdb.org/en/stable/api/server/authn.html#proxy-authentication) so CouchDB uses the OAuth service to do auth. This is less monolithic but I worry the migration would be a pain - it would probably completely break the user management app for example.

I recommend asking in the [CouchDB Slack channel](https://join.slack.com/t/couchdb/shared_invite/zt-fa9zim0j-H04m4o_KcLdWeOxEAcwM8g) - someone must have done this before...

**Comment by jkuester** [comment]
https://github.com/apache/couchdb/issues/1018 I think Couch proxy auth is absolutely the "correct" way to go here.  In an ideal world we could have a de-coupled authentication server that is used by both the CHT api and Couch to manage users. Things could get really interesting at that point since the integration points would just be via standard OAuth/JWT interactions so, in theory, you could even use a 3rd party auth server. We could have a simple auth/user setup for smaller partners, and anyone who needed something more sophisticated could integrate with something like KeyCloak (or whatever auth server they already have...).

But that is all dreams at this point (that may or may not ever be useful to anyone...).  Based on the conversations we have had so far, I think our primary goal with this MVP is to implement OAuth while still using Couch as the source-of-truth for users (change as little as possible).  It is possible we will determine that this approach is not feasible, but at this point I don't think we have a good enough idea of what it might look like to make that determination.

**Comment by jkuester** [comment]
Also, started [Couch Slack Discussion here](https://couchdb.slack.com/archives/C49LEE7NW/p1707491311095569)...

**Comment by jkuester** [comment]
Update from our design discussion today: after further evaluation of the code, we determined that *if* we want to continue using Couch as our identity provider and *if* we want to use the CHT api as our authorization server, then the [[express-oauth-server](https://github.com/node-oauth/express-oauth-server)](https://github.com/node-oauth/express-oauth-server) package seems to be the best way forwards.  It should integrate easily as Express middleware while still providing flexibility for connecting everything to Couch users/sessions. The only caveat is that it is not clear what support there is for OIDC in this package.  @sugat009 is going to do more research here.  It is also clear that this approach would probably require implementing some kind of in-memory caching to avoid extra round-trips to Couch for auth flows.

That being said, from the CouchDB Slack thread it is pretty clear that using CouchDB as the identity provider/auth server in an OAuth context is not the typical setup. JWT/Proxy auth for Couch via an external identity provider/auth server seems more typical/recommended when doing OAuth.  One remaining wildcard here is the [[couch-auth](https://github.com/perfood/couch-auth)](https://github.com/perfood/couch-auth) package that was also mentioned.  This package seems to allow bridging users from OAuth to CouchDB, but it maybe expects an external authentication server (e.g. your application could let users login with their Google accounts (via Google's OAuth), and then couch-auth would do the mapping of those Google accounts to Couch users....).  I am going to be looking more into this package.

The final design thread I would like to pursue before trying to make a decision here is to map out what it could look like to implement JWT/Proxy auth for Couch along with an external identity provider (and maybe even an external authentication server).  This seems like a pretty drastic step just to implement OAuth, but there could be benefits. It may allow us to have a more drop-in approach to auth without having to do a bunch of custom logic to map things to Couch sessions.  Also, decoupling identity/authentication from Couch could provide a lot more flexibility for the CHT ecosystem in the future....  Seems worth at least having this proposal on the table.

Once we have satisfactorily investigated these remaining open questions, we will put together a more coherent design proposal laying out the different options as we see them, so Gareth (et al) can make the final decision on how to proceed!

**Comment by jkuester** [comment]
Okay, so my hot take is that implementing OAuth in the CHT for the sole purpose of enabling 3rd Party integrations (while still supporting cookie/basic-auth in parallel for the long-term) is not going to be worth the effort to implement (or the added cost to continue to maintain the OAuth logic in the code base).  I think we need to decide if we are interested in (eventually) moving our whole setup from cookie/basic-auth to [OAuth/API-keys](https://stackoverflow.blog/2021/10/06/best-practices-for-authentication-and-authorization-for-rest-apis/) for greater security, flexibility, and integrations. Of course, this move should be done incrementally, but I don't think we want to chain ourselves to OAuth at all if it is just for 3rd Party integrations.....

If we actually do want to implement in OAuth in the CHT, here are the 4 alternative approaches I have come up with for doing this!  (Well the first is the "do nothing" case, so it does not really count.)  These describe the target "end-state" that we would shoot for. It is likely that we could actually ship more incremental changes. 

<details><summary><h2>1. Current</h2></summary>
<p>

![image](https://github.com/medic/cht-core/assets/8441903/392cfe10-0c14-4aed-ae66-2853f6a528fc)

Couch is the Authorization Server and the Identity Provider.

#### Pros

- No changes required.
- Simple - Auth delegated to Couch

#### Cons

- No way to give *scoped* or *revocable* access to resources for a particular user.
- No way for 3rd party client to authenticate without asking user for username/password.
- No way to integrate with 3rd party Identity Providers.

</p>
</details> 
<details><summary><h2>2. Wrapper Auth</h2></summary>
<p>

![image](https://github.com/medic/cht-core/assets/8441903/c3b76f2d-90da-4db6-a07c-e0e534b14518)

CHT api is Authorization Server (supporting OAuth/OIDC), but really just acts as a wrapper around Couch users. Couch remains the Identity Provider and source of truth for user authentication.

#### Pros

- No user migration required.
- Simple to deploy since there are not any additional moving parts or changes to Couch.
- OAuth/OIDC flow for 3rd party authentication.

#### Cons

- Would still need to have mapping between Auth Tokens and Couch session cookies.
	- Using the Couch session cookie as the Auth Token would mean not having to maintain a stateful storage of the mapping, but it would also mean that these tokens would not actually be  *scoped* or *revocable*.
- Statefully storing a mapping of Auth Tokens to Couch sessions would probably require some kind of in-memory cache to avoid performance issues.
- No way to integrate with 3rd party Identity Providers.

</p>
</details> 
<details><summary><h2>3. Circular Auth</h2></summary>
<p>

![image](https://github.com/medic/cht-core/assets/8441903/16ebaacd-ac03-46c8-9e8b-1b16390a2a10)

CHT api is Authorization Server and (for all intents and purposes) the Identity Provider, but it still uses Couch as its *user storage* mechanism. Couch is configured to use JWT authentication and the CHT api is responsible for creating the JWTs. (The "circle" here is just that the api would use data from Couch to provision JWTs and then use the JWTs to authorize access to Couch.)

#### Pros:

- Seems possible that this could be done without actually doing any significant user migration.  The api could just store/lookup user data in the `_users` db.
- Could possibly just re-use the Couch JWT as the Auth Token, simplifying the amount of moving parts here.
- OAuth/OIDC flow for 3rd party authentication.
#### Cons:
- Lots of custom logic to add in api to do all the things.
	- Would need a Couch user configured for the API to use for getting user data from Couch.
- No way to integrate with 3rd party Identity Providers.

</p>
</details> 
<details><summary><h2>4. Decoupled IAM</h2></summary>
<p>

![image](https://github.com/medic/cht-core/assets/8441903/e4eeeea1-c159-44fc-b197-6155db4e31b0)

Delegate everything to a separate Authentication Server and Identity Provider (either roll-our-own, integrate a 3rd party, or say "bring your own").  Clients authenticate against the auth server (via OAuth) and then can call the CHT api with their auth tokens.

#### Pros:

- All interaction with the IAM system would be via OAuth/OIDC. This loose coupling on standard interfaces would conceivably allow for integrations with a range of 3rd party IAM solutions.
	- For example, a partner could integrate the CHT with their existing IAM so that CHT user accounts are integrated with other partner systems (e.g. email accounts) and everything can be centrally managed (and/or could have federated user identity....).
- No extra logic would have to be included in the CHT code.  Instead, the api *authorization* logic would be very close to how things are currently (and we would not need any *authentication* logic).
- OAuth/OIDC flow for 3rd party authentication.

#### Cons:

- Would have to migrate users from Couch to IAM system. This could range from tricky to just impossible...
- More moving parts! Another (or several) point of failure for things to go wrong.

</p>
</details>

**Comment by mrjones-plip** [comment]
Pulling in the update from slack on this to shed more light on why we're closing it.

I said:

> @jkuester - You've done a really impressive, deep dive into how all the pieces fit together to assess [which ideas](https://github.com/medic/cht-core/issues/8849#issuecomment-1949200134) will best work and HOW they all work in comparison.  Thank you! I really enjoyed our call earlier today
>
> After chatting we decided: Allies is NOT implementing OAuth2. It is too complicated and provides no real benefit to 3rd party integrations like the User Management Tool. The only reason we'd consider doing work on this now is if CHT as a whole wanted to provide this feature as a heavy lift, long term, complex effort (read Q1 + Q2)


And @jkuester replied:

> Just a couple wording clarifications for the benefit of other readers (based on what we already discussed)!   :sweat_smile:
> 
>> provides no real benefit to 3rd party integrations like the User Management Tool
>
> OAuth would not enable any feature that is currently missing from cht-user-management. There is already a "workaround" for authentication (in the form of just accepting username/password).  It is appalling from a security perspective, but it "works".  SSO (which OAuth does provide) would be a benefit to all 3rd party apps in terms of ease of integration and user security, but neither of these are at the top of the current priorities for cht-user-managment.
>
>> The only reason we'd consider doing work on this now is if CHT as a whole wanted to provide this feature
>
> Meaning that we felt it was important right now for the CHT to support security best practices, allowing for access to resources in a way that is scope-able and revocable (currently user's can only give all-or-nothing access to their account and the only way to revoke access is to change the password). If this was not at least a long term goal, then OAuth is just a really over-architected way of doing SSO...

**Comment by jkuester** [comment]
@mrjones-plip Not sure we actually want to _close_ this issue.  IMHO we decided we don't want to do this _now_, but we did not decide that we _never_ want to support OAuth.  Seems fine to leave this issue open to track future interest in this feature...

**Comment by mrjones-plip** [comment]
ah - thank you @jkuester - excellent point about keeping it open as to track interest. I'll open this and the [OIDC one](https://github.com/medic/cht-core/issues/8850) too.

**Comment by vikrantwiz02** [comment]
Hey @jkuester @mrjones-plip, I spent some time looking at the complexity concerns that led to closing this, and I think there’s actually a clean way to solve them.

The sticking point in Wrapper Auth was always the stateful mapping between OAuth tokens and CouchDB sessions, you'd need a cache, maintenance overhead, the whole thing. But if the access token is a **stateless JWT** signed with CouchDB's own `couch_httpd_auth/secret` (which `sso-login.js` already reads), that mapping disappears entirely. Every Bearer token request validates locally via HMAC —> no DB roundtrip, no cache.

On the CouchDB side, nothing changes. This is explicitly not Circular Auth, we're not reconfiguring Couch to accept JWTs. The API validates the token, sets `req.userCtx`, and the rest of the existing auth middleware runs exactly as-is. Cookie and Basic auth paths are untouched too.

Concretely it's about 5 files: a new service for client registration + JWT sign/verify, a controller for the authorize/token endpoints, a consent page template, one new branch in `auth.getUserCtx`, and route registrations. Auth codes get deleted on use so CouchDB's MVCC handles concurrent replay natively —> no soft-delete flag, no race condition.

MVP scope would be Authorization Code flow only (confidential clients), single `cht` scope that just inherits the user's existing roles, and an admin API for registering clients. No refresh tokens, no PKCE, no introspection —> all explicitly deferred.

Curious if this approach works better for you...?

**Comment by jkuester** [comment]
>  But if the access token is a stateless JWT signed with CouchDB's own couch_httpd_auth/secret (which sso-login.js already reads), that mapping disappears entirely.

🤔 the signed JWT solves some problems here for sure, but I am not sure there is a simple way to make individual JWTs "scoped or revocable."  IMHO, whatever our next-gen auth system is, it needs to support these concepts. (Related to https://github.com/medic/cht-core/issues/10342).  

And honestly, I am not sure how much traction there is currently for adding support to the CHT for being an _OIDC Provider_. Our existing SSO functionality is basically `4. Decoupled IAM` and that provides a lot of flexibility for different workflows, etc. I don't know of anything blocking us from proceeding with this issue in the future, but right now I do not know of anyone asking for it.

**Comment by vikrantwiz02** [comment]
@jkuester, fair enough on both points. The revocation/scoping gap is real, a short-lived JWT buys you some protection but it's not a substitute for actual revocability, and shipping half of that story isn't great.

Read through #10342 and I can see why you're holding it back too, the same permissions scoping problem blocks both, and that's clearly a design conversation that needs to happen before any of this makes sense to code.

I'll keep an eye on how that design thinking evolves. Thanks for taking the time to walk through it.
